torch/csrc/Dtype.cpp

```diff
@@ -1,134 +1,115 @@
-#include <torch/csrc/utils/tensor_dtypes.h>
-#include <torch/csrc/utils/tensor_types.h>
+#include <torch/csrc/utils/pybind.h>
 
-#include <torch/csrc/Exceptions.h>
+#include <sstream>
 
-PyObject * THPDtype_New(at::ScalarType scalar_type, const std::string& name)
-{
-  AT_ASSERT(name.length() < DTYPE_NAME_LEN);
-  auto type = (PyTypeObject*)&THPDtypeType;
-  auto self = THPObjectPtr{type->tp_alloc(type, 0)};
-  if (!self) throw python_error();
-  auto self_ = reinterpret_cast<THPDtype*>(self.get());
-  self_->scalar_type = scalar_type;
-  std::strncpy(self_->name, name.c_str(), DTYPE_NAME_LEN);
-  return self.release();
-}
+namespace torch {
+using at::ScalarType;
 
-PyObject *THPDtype_is_floating_point(THPDtype *self, PyObject *noargs)
-{
-  if (at::isFloatingType(self->scalar_type)) {
-    Py_RETURN_TRUE;
-  } else {
-    Py_RETURN_FALSE;
-  }
-}
+namespace {
+#define PY_DTYPE(scalar_type, ...) { true, ScalarType::scalar_type, __VA_ARGS__ }
 
-PyObject *THPDtype_is_complex(THPDtype *self, PyObject *noargs)
-{
-  if (at::isComplexType(self->scalar_type)) {
-    Py_RETURN_TRUE;
-  } else {
-    Py_RETURN_FALSE;
-  }
+constexpr size_t NUM_DTYPES = static_cast<size_t>(ScalarType::NumOptions);
+// Must be in exactly the same order as ScalarType definitions
+constexpr PyDtype dtype_registry[NUM_DTYPES] = {
+  PY_DTYPE(Byte, "uint8", ""),
+  PY_DTYPE(Char, "int8", ""),
+  PY_DTYPE(Short, "int16", "short"),
+  PY_DTYPE(Int, "int32", "int"),
+  PY_DTYPE(Long, "int64", "long"),
+  PY_DTYPE(Half, "float16", "half"),
+  PY_DTYPE(Float, "float32", "float"),
+  PY_DTYPE(Double, "float64", "double"),
+  PY_DTYPE(ComplexHalf, "complex32", ""),
+  PY_DTYPE(ComplexFloat, "complex64", "cfloat"),
+  PY_DTYPE(ComplexDouble, "complex128", "cdouble"),
+  PY_DTYPE(Bool, "bool", ""),
+  PY_DTYPE(QInt8, "qint8", ""),
+  PY_DTYPE(QUInt8,"quint8", ""),
+  PY_DTYPE(QInt32, "qint32", ""),
+  PY_DTYPE(BFloat16, "bfloat16", ""),
+  PY_DTYPE(QUInt4x2, "quint4x2", ""),
+};
+
+c10::string_view dtypeName(const PyDtype& dtype) {
+  return dtype.primary_name;
 }
 
-PyObject *THPDtype_is_signed(THPDtype *self, PyObject *noargs)
-{
-  HANDLE_TH_ERRORS
-  if (at::isSignedType(self->scalar_type)) {
-    Py_RETURN_TRUE;
-  } else {
-    Py_RETURN_FALSE;
+} // namespace
+
+const PyDtype& getPyDtype(ScalarType scalar_type) {
+  const PyDtype& dtype = dtype_registry[static_cast<size_t>(scalar_type)];
+  if (!dtype.defined) {
+    throw std::invalid_argument("unsupported ScalarType");
   }
-  END_HANDLE_TH_ERRORS
+  return dtype;
 }
 
-PyObject *THPDtype_reduce(PyObject *_self, PyObject *noargs)
-{
-  /*
-  * For singletons, a string is returned. The string should be interpreted
-  * as the name of a global variable.
-  */
-  auto self = (THPDtype*)_self;
-  return THPUtils_packString(self->name);
-}
+void initDtypeBindings(PyObject* module) {
+#define ASSERT_PY_DTYPE(_1, type) \
+  static_assert( \
+    dtype_registry[static_cast<size_t>(ScalarType::type)].defined, \
```

Comment:
just thinking- do we technically need to store extra space for a `defined` flag directly in the struct? I guess the question is "if someone fiddles around with the code for `dtype_registry` in this file, would the second static assert be enough to catch it"?

The old code would construct  the `dtype_registry` table at runtime, so its notion of "defined" was "not all 0 bits", which is never the case here since you made the table `constexpr`. So at the very least, I don't think any of your runtime checks for `if (!dtype.defined)` are necessary, thanks to your static assert here.


The old code didn't have it, and its notion of "defined" was "not all 0 bits". The old code would construct that `dtype_registry` at runtime, but you construct it at compile time